android.database.sqlite.SQLiteException: duplicate column name: wiki_data_id (code 1): , while compiling: ALTER TABLE additive_name ADD COLUMN 'wiki_data_id' TEXT NOT NULL DEFAULT '';

04-09 22:19:57.542 1551-1551/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.openpetfoodfacts.scanner, PID: 1551
    java.lang.RuntimeException: Unable to create application openfoodfacts.github.scrachx.openfood.views.OFFApplication: android.database.sqlite.SQLiteException: duplicate column name: wiki_data_id (code 1): , while compiling: ALTER TABLE additive_name ADD COLUMN 'wiki_data_id' TEXT NOT NULL DEFAULT '';
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5825)
        at android.app.ActivityThread.-wrap1(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1666)
        at android.os.Handler.dispatchMessage(Handler.java:105)
        at android.os.Looper.loop(Looper.java:251)
        at android.app.ActivityThread.main(ActivityThread.java:6572)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)
     Caused by: android.database.sqlite.SQLiteException: duplicate column name: wiki_data_id (code 1): , while compiling: ALTER TABLE additive_name ADD COLUMN 'wiki_data_id' TEXT NOT NULL DEFAULT '';
        at android.database.sqlite.SQLiteConnection.nativePrepareStatement(Native Method)
        at android.database.sqlite.SQLiteConnection.acquirePreparedStatement(SQLiteConnection.java:895)
        at android.database.sqlite.SQLiteConnection.prepare(SQLiteConnection.java:506)
        at android.database.sqlite.SQLiteSession.prepare(SQLiteSession.java:588)
        at android.database.sqlite.SQLiteProgram.<init>(SQLiteProgram.java:58)
        at android.database.sqlite.SQLiteStatement.<init>(SQLiteStatement.java:31)
        at android.database.sqlite.SQLiteDatabase.executeSql(SQLiteDatabase.java:1677)
        at android.database.sqlite.SQLiteDatabase.execSQL(SQLiteDatabase.java:1608)
        at org.b.a.b.f.a(Unknown Source:2)
        at openfoodfacts.github.scrachx.openfood.models.DatabaseHelper.upgrade(Unknown Source:29)
        at openfoodfacts.github.scrachx.openfood.models.DatabaseHelper.onUpgrade(Unknown Source:34)
        at org.b.a.b.b.onUpgrade(Unknown Source:4)
        at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:299)
        at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:194)
        at org.b.a.b.b.getWritableDb(Unknown Source:0)
        at openfoodfacts.github.scrachx.openfood.views.OFFApplication.onCreate(Unknown Source:55)
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1119)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5822)
        at android.app.ActivityThread.-wrap1(Unknown Source:0) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1666) 
        at android.os.Handler.dispatchMessage(Handler.java:105) 
        at android.os.Looper.loop(Looper.java:251) 
        at android.app.ActivityThread.main(ActivityThread.java:6572) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767) 

--- a/app/src/main/java/openfoodfacts/github/scrachx/openfood/models/DatabaseHelper.java
+++ b/app/src/main/java/openfoodfacts/github/scrachx/openfood/models/DatabaseHelper.java
@@ -1,12 +1,11 @@
 package openfoodfacts.github.scrachx.openfood.models;
 
 import android.content.Context;
+import android.database.Cursor;
 import android.database.sqlite.SQLiteDatabase;
 import android.util.Log;
 
 import org.greenrobot.greendao.database.Database;
-
-import static openfoodfacts.github.scrachx.openfood.models.DaoMaster.dropAllTables;
 
 public class DatabaseHelper extends DaoMaster.OpenHelper {
 
@@ -80,26 +79,33 @@ public class DatabaseHelper extends DaoMaster.OpenHelper {
                 break;
             }
             case 7: {
-                db.execSQL("ALTER TABLE additive_name ADD COLUMN 'wiki_data_id' TEXT NOT NULL DEFAULT '';");
-                db.execSQL("ALTER TABLE additive_name ADD COLUMN 'is_wiki_data_id_present' INTEGER NOT NULL DEFAULT '';");
-
-                db.execSQL("ALTER TABLE additive ADD COLUMN 'wiki_data_id' TEXT NOT NULL DEFAULT '';");
-                db.execSQL("ALTER TABLE additive ADD COLUMN 'is_wiki_data_id_present' INTEGER NOT NULL DEFAULT '';");
-
-                db.execSQL("ALTER TABLE category_name ADD COLUMN 'wiki_data_id' TEXT NOT NULL DEFAULT '';");
-                db.execSQL("ALTER TABLE category_name ADD COLUMN 'is_wiki_data_id_present' INTEGER NOT NULL DEFAULT '';");
-
-                db.execSQL("ALTER TABLE category ADD COLUMN 'wiki_data_id' TEXT NOT NULL DEFAULT '';");
-                db.execSQL("ALTER TABLE category ADD COLUMN 'is_wiki_data_id_present' INTEGER NOT NULL DEFAULT '';");
-
-                db.execSQL("ALTER TABLE label_name ADD COLUMN 'wiki_data_id' TEXT NOT NULL DEFAULT '';");
-                db.execSQL("ALTER TABLE label_name ADD COLUMN 'is_wiki_data_id_present' INTEGER NOT NULL DEFAULT '';");
-
-                db.execSQL("ALTER TABLE label ADD COLUMN 'wiki_data_id' TEXT NOT NULL DEFAULT '';");
-                db.execSQL("ALTER TABLE label ADD COLUMN 'is_wiki_data_id_present' INTEGER NOT NULL DEFAULT '';");
+                String newColumns[] = new String[]{"wiki_data_id", "is_wiki_data_id_present"};
+                String updatedTables[] = new String[]{"additive_name", "additive", "category_name", "category", "label_name", "label"};
+                for (String table : updatedTables) {
+                    for (String column : newColumns) {
+                        if (isFieldExist(db, table, column)) {
+                            db.execSQL(String.format("ALTER TABLE %s ADD COLUMN '%s' TEXT NOT NULL DEFAULT '';", table, column));
+                        }
+                    }
+                }
 
                 break;
             }
         }
+    }
+
+    private boolean isFieldExist(Database db, String tableName, String fieldName) {
+        boolean isExist = false;
+        String query = String.format("PRAGMA table_info(%s)", tableName);
+        Cursor res = db.rawQuery(query, null);
+        res.moveToFirst();
+        do {
+            String currentColumn = res.getString(1);
+            if (currentColumn.equals(fieldName)) {
+                isExist = true;
+            }
+        } while (res.moveToNext());
+
+        return isExist;
     }
 }

